lib/find-node-directory.js

```diff
@@ -0,0 +1,40 @@
+var path = require('path')
+  , log = require('npmlog')
+
+function findNodeDirectory() {
+  // Have a look to see what is above us, to try and work out where we are
+  npm_parent_directory = path.join(__dirname, '../../../..')
+  log.verbose('node-gyp root', 'npm_parent_directory is ' + path.basename(npm_parent_directory))
+  node_root_dir = ""
+
+  log.verbose('node-gyp root', 'Finding node root directory')
+  if (path.basename(npm_parent_directory) == 'lib') {
+    // We are in a node install directory
+    node_root_dir = path.join(npm_parent_directory, '../..')
+    log.verbose('node-gyp root', 'in install directory, root = ' + node_root_dir)
+  } else if (path.basename(npm_parent_directory) == 'node_modules') {
+    // We are in a node build directory, where npm lives under lib/
+    node_root_dir = path.join(npm_parent_directory, '../../include/node')
+    log.verbose('node-gyp root', 'in build directory, root = ' + node_root_dir)
+  } else {
+    // We don't know where we are, try working it out from the location
+    // of the node binary
+    var node_dir = path.dirname(process.execPath)
+    var directory_up = path.basename(node_dir)
+    if (directory_up == 'bin') {
+      node_root_dir  = path.join(node_dir, '..')
+    } else if (directory_up == 'Release' || directory_up == 'Debug') {
+      // If we are a recently built node, and the directory structure is that of a repository
+      // If we are on Windows then we only need to go one level up, everything else, two
+      if (process.platform == 'win32') {
+        node_root_dir = path.join(node_dir, '..')
+      } else {
+        node_root_dir = path.join(node_dir, '../..')
+      }
+    }
+    // Else return the default blank, "".
+  }
+  return node_root_dir
+}
```

Comment:
Can you add a few regression tests for this function?  I'd make it so that you can plug in a mock `__dirname` and `process` object.